fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/services/MembershipServiceImpl.java

```diff
@@ -0,0 +1,398 @@
+                if (memberDeleted) {
+                    memberCreated = ((Tombstone) member).getDeletedObject().getCreatedDate();
+                } else {
+                    memberCreated = member.getCreatedDate();
+                }
+                final var memberModified = member.getLastModifiedDate();
+                final var memberEnd = memberDeleted ? memberModified : NO_END_INSTANT;
+
+                // Reduce timeline to just states in effect after the member was created
+                var timelineStream = propertyTimeline.stream()
+                        .filter(e -> e.endDatetime.compareTo(memberCreated) > 0);
+                // If the member was deleted, then reduce timeline to states before the deletion
+                if (memberDeleted) {
+                    timelineStream = timelineStream.filter(e -> e.mementoDatetime.compareTo(memberModified) < 0);
+                }
+                // Index each addition or change to the membership generated by this member
+                timelineStream.forEach(e -> {
+                    // Start time of the membership is the later of member creation or membership resc memento time
+                    indexManager.addMembership(txId, containerId,
+                            generateMembershipTriple(e.membershipResource,
+                                    memberNode, e.hasMemberRelation, e.isMemberOfRelation),
+                            instantMax(memberCreated, e.mementoDatetime),
+                            instantMin(memberEnd, e.endDatetime));
+                });
+            });
+        }
+    }
+
+    private Instant instantMax(final Instant first, final Instant second) {
+        if (first.isAfter(second)) {
+            return first;
+        } else {
+            return second;
+        }
+    }
+
+    private Instant instantMin(final Instant first, final Instant second) {
+        if (first.isBefore(second)) {
+            return first;
+        } else {
+            return second;
+        }
+    }
+
+    /**
+     * Creates a timeline of states for a DirectContainer, tracking changes to its
+     * properties that impact membership.
+     * @param fedoraResc resource subject of the timeline
+     * @return timeline
+     */
+    private List<DirectContainerProperties> makePropertyTimeline(final FedoraResource fedoraResc) {
+        final var entryList = fedoraResc.getTimeMap().getChildren()
+                .sorted((m1, m2) -> m1.getFedoraId().getFullId().compareTo(m1.getFedoraId().getFullId()))
+                .map(memento -> new DirectContainerProperties(memento))
+                .collect(Collectors.toList());
+
+        // Reduce timeline to entries where significant properties change
+        final var changeEntries = new ArrayList<DirectContainerProperties>();
+        var curr = entryList.get(0);
+        changeEntries.add(curr);
+        int i = 1;
+        while (i < entryList.size()) {
```

Comment:
Why are you using a while loop rather than a for loop?

Reply:
Hmm, I'd like to think I had a reason at some point, but doesn't seem like there is one, I'll switch it.